### PR TITLE
Development

### PR DESCRIPTION
**Pull request by BjornMelgaard**

### CODE REVIEW

File: app/decorators/shopper/checkout_page/delivery_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+require_relative '../concerns/view_helpers'
```
Comment (kirillweb) [inline app/decorators/shopper/checkout_page/delivery_decorator.rb]: Better to require all into [shopper.rb](https://github.com/BjornMelgaard/rails-bookstore-engine/blob/development/lib/shopper.rb)

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+remove-db:
+	rails app:db:environment:set RAILS_ENV=$(ENV)
+	rails app:db:drop RAILS_ENV=$(ENV)
+
+create-db:
+	rails app:db:create RAILS_ENV=$(ENV)
+	rails app:db:migrate RAILS_ENV=$(ENV)
+
+dummy-server:
+	(cd spec/dummy && rails server)
```
Comment (kirillweb) [inline Makefile]: For what purposes did you create the makefile?)

File: app/commands/shopper/checkout_page/add_checkout_addresses.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+      def call
+        @use_billing = use_billing?
+
+        set_address(:billing)
+        set_address(:shipping) unless @use_billing
+
+        unless [@billing, @shipping].compact.all?(&:valid?)
+          broadcast(:invalid, @order, @billing, @shipping, @use_billing)
+          return
```
Comment (kirillweb) [inline app/commands/shopper/checkout_page/add_checkout_addresses.rb]:
```ruby 
return broadcast(:invalid, @order, @billing, @shipping, @use_billing) 
```

File: app/controllers/shopper/cart_controller.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        end
+        on(:ok) { redirect_to cart_path, flash: { success: 'Cart was updated successfully'  } }
+      end
+    end
+
+    def add_product
+      CartPage::AddProduct.call(params) do
+        on(:invalid_product) do
+          redirect_to main_app.root_path, flash: { error: 'Invalid product' }, js: true
+          return
```
Comment (kirillweb) [inline app/controllers/shopper/cart_controller.rb]: why do you need it?

File: app/commands/shopper/cart_page/update_order_items.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+      private
+
+      def update
+        @order.order_items.update(iparams.keys, iparams.values)
+      end
+
+      def iparams
+        @iparams = @params.permit(items: [:quantity]).require(:items)
+      rescue
```
Comment (kirillweb) [inline app/commands/shopper/cart_page/update_order_items.rb]: looks strange, why do you handle exception here? what can happen in the line above?

File: app/controllers/shopper/cart_controller.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  class CartController < ShopperController
+    respond_to :js, only: [:add_product]
+
+    def show
+      present CartPage::CartPresenter.new(current_order)
+    end
+
+    def update
+      CartPage::UpdateCart.call(params, current_order) do
+        on(:invalid_coupon)  { flash[:error] = 'Invalid coupon code' }
```
Comment (kirillweb) [inline app/controllers/shopper/cart_controller.rb]: Better to move all text messages into i18n

File: spec/support/show_me_the_cookies.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+RSpec.configure do |config|
+  config.include ShowMeTheCookies, :type => :feature
```
Comment (kirillweb) [inline spec/support/show_me_the_cookies.rb]:
```ruby
config.include ShowMeTheCookies, type: :feature
```

File: spec/support/capybara.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+# for email testing
+Capybara.server_port = 3001
+Capybara.app_host = 'http://localhost:3001'
+
+# Capybara::Webkit.configure do |config|
+#   config.block_unknown_urls
+#   # config.allow_url('w2.yotpo.com')
+#   # config.allow_url('staticw2.yotpo.com')
+#   # config.allow_url('fonts.googleapis.com')
+# end
```
Comment (kirillweb) [inline spec/support/capybara.rb]: ?

File: app/commands/shopper/checkout_page/place_order.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        @order.customer = current_customer
+        @order.save
+      end
+
+      def send_email
+        Shopper::CheckoutMailer.complete(current_customer, @order).deliver_later
+      end
+    end
+  end
+end
```
Comment (kirillweb) [inline app/commands/shopper/checkout_page/place_order.rb]: tests?)

File: app/commands/shopper/checkout_page/add_checkout_addresses.rb (hunk 3 above)
Comment (BjornMelgaard) [inline app/commands/shopper/checkout_page/add_checkout_addresses.rb]: human can keep in head 7+-2 parameters simultaneously, everything is fine)

File: Makefile (hunk 2 above)
Comment (BjornMelgaard) [inline Makefile]: In order to not keep that in head

File: app/controllers/shopper/cart_controller.rb (hunk 4 above)
Comment (BjornMelgaard) [inline app/controllers/shopper/cart_controller.rb]: invalid_product? for intruders

File: spec/support/capybara.rb (hunk 8 above)
Comment (BjornMelgaard) [inline spec/support/capybara.rb]: )

File: app/controllers/shopper/cart_controller.rb (hunk 4 above)
Comment (kirillweb) [inline app/controllers/shopper/cart_controller.rb]: return - why do need here?

Comment (kirillweb) [inline app/controllers/shopper/cart_controller.rb]: just redirect?

Comment (BjornMelgaard) [inline app/controllers/shopper/cart_controller.rb]: it will call render implicitly

Comment (BjornMelgaard) [inline app/controllers/shopper/cart_controller.rb]: Had changed to without return, it kind of works

**Pull request merged**